doc/pipeline_spec.md

```diff
@@ -43,17 +43,92 @@
 
-`inputs.method` specifies how a repo will be partitioned among parallel containers, and whether the entire repo or just the new commit is used as the input.
+`inputs.method` specifies two different properties:
+- Partition unit: How input data  will be partitioned across parallel containers.
+- Incrementality: Whether the entire all of the data or just the new data (diff) is processed. 
+ 
+The next section explains input methods in detail.
 
-You may specify a method using either an alias or a JSON object.  We support four aliases that represent the four commonly used strategies:
+### Pipeline Input Methods
 
-* map: each job sees a part of the new commit; files may be partitioned.
-* reduce: each job sees a part of the entire repo; files are not partitioned
-* incremental_reduce: each job sees a part of the new commit; files are not partitioned.
-* global: each job sees the entire repo
+For each pipeline input, you may specify a "method".  A method dictates exactly what happens in the pipeline when a commit comes into the input repo.
 
-If a method is not specified, the "map" method is used by default.
+A method consists of two properties: partition unit and incrementality.
 
+#### Partition Unit
+Partition unit specifies the granularity at which input data is parallelized across containers.  
+
+* `Blocks`: different blocks of the same file may be parelleized across containers.
+* `Top-level objects`: the files and/or directories residing under the root directory (/) must be grouped together.  For instance, if you have four files in a directory structure like: 
+
+```
+/foo 
+/bar
+/buzz
+   /a
+   /b
+```
+then there are only three top-level objects, `/foo`, `/bar`, and `/buzz`, each of which will remain grouped in the same container. 
+* `Repo`: the entire repo.  In this case, the input won't be partitioned at all. 
+
+#### Incrementality
+
+Incrementality is a boolean flag that describes what data needs to be available when a new commit is made on an input repo. Namely, do you want to process only the new data in that commmit (the diff) or does all of the data need to be reprocessed?
+
+For instance, if you have a repo with the file `/foo` in commit 1 and file `/bar` in commit 2, then:
+
+* If the input is incremental, the first job sees file `/foo` and the second job sees file `/bar`.
+* If the input is nonincremental, the first job sees file `/foo` and the second job sees file `/foo` and file `/bar`.
+
+For convenience, we have defined aliases for the four most commonly used input methods: map, reduce, incremental-reduce, and global.  They are defined below:
+
+|                | Block |  Top-level Objects |  Repo  |
+|----------------|-------|--------------------|--------|
+|   Incremental  |  map  | incremental-reduce |        |
+| Nonincremental |       |       reduce       | global |
+
+#### Defaults
+If no method is specified, the `map` method (Block + Incremental) is used by default.
```

Comment:
Makes more sense to me that you need to explicitly specify both, otherwise we error.  Most users are expected to use aliases anyways.